formula/src/main/java/com/instacart/formula/internal/FormulaManagerImpl.kt

```diff
@@ -39,30 +39,53 @@ internal class FormulaManagerImpl<Input, State, Output>(
 
     var transitionID: Long = 0
     var terminated = false
+    private var isRunning: Boolean = false
```

Comment:
We keep track if we are within the `run` function. If so, we can perform a bunch of logic inline without the need for a full re-run.